internal/dispatch/graph/reachableresources_test.go

```diff
@@ -719,3 +723,107 @@ func TestReachableResourcesMultipleEntrypointEarlyCancel(t *testing.T) {
+	defer goleak.VerifyNone(t, goleakIgnores...)
+
+	rawDS, err := memdb.NewMemdbDatastore(0, 0, memdb.DisableGC)
+	require.NoError(t, err)
+
+	testRels := make([]*core.RelationTuple, 0)
+	for i := 0; i < 410; i++ {
+		testRels = append(testRels, tuple.MustParse(fmt.Sprintf("resource:res%d#viewer@user:tom", i)))
+	}
+
+	ds, revision := testfixtures.DatastoreFromSchemaAndTestRelationships(
+		rawDS,
+		`
+			definition user {}
+
+			definition resource {
+				relation viewer: user
+				permission view = viewer
+			}
+		`,
+		testRels,
+		require.New(t),
+	)
+	dispatcher := NewLocalOnlyDispatcher(2)
+
+	ctx := log.Logger.WithContext(datastoremw.ContextWithHandle(context.Background()))
+	require.NoError(t, datastoremw.SetInContext(ctx, ds))
+
+	// Dispatch reachable resources but stop reading after the first 1000 results.
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	stream := dispatch.NewCollectingDispatchStream[*v1.DispatchReachableResourcesResponse](ctxWithCancel)
+	err = dispatcher.DispatchReachableResources(&v1.DispatchReachableResourcesRequest{
+		ResourceRelation: RR("resource", "view"),
+		SubjectRelation: &core.RelationReference{
+			Namespace: "user",
+			Relation:  "...",
+		},
+		SubjectIds: []string{"tom"},
+		Metadata: &v1.ResolverMeta{
+			AtRevision:     revision.String(),
+			DepthRemaining: 50,
+		},
+	}, stream)
+	require.NoError(t, err)
+	defer cancel()
+
+	foundResources := util.NewSet[string]()
+	var cursor *v1.Cursor
+
+	for index, result := range stream.Results() {
+		for _, resource := range result.Resources {
+			require.True(t, foundResources.Add(resource.ResourceId))
+		}
+
+		// Break on the second result.
+		if index == 1 {
+			cursor = result.AfterResponseCursor
+			cancel()
+			break
+		}
+	}
+
+	// Ensure we've found a cursor and that we got 200 resources back in the first + second result.
+	require.NotNil(t, cursor, "got no cursor and %d results", foundResources.Len())
+	require.Equal(t, foundResources.Len(), 200)
+
+	// Call reachable resources again with the cursor, which should continue in the second result
+	// and then move forward from there.
+	stream2 := dispatch.NewCollectingDispatchStream[*v1.DispatchReachableResourcesResponse](ctx)
+	err = dispatcher.DispatchReachableResources(&v1.DispatchReachableResourcesRequest{
+		ResourceRelation: RR("resource", "view"),
+		SubjectRelation: &core.RelationReference{
+			Namespace: "user",
+			Relation:  "...",
+		},
+		SubjectIds: []string{"tom"},
+		Metadata: &v1.ResolverMeta{
+			AtRevision:     revision.String(),
+			DepthRemaining: 50,
+		},
+		OptionalCursor: cursor,
+	}, stream2)
+	require.NoError(t, err)
+
+	count := 0
+	overlappingCount := 0
+	for _, result := range stream2.Results() {
+		for _, resource := range result.Resources {
+			count++
+			if !foundResources.Add(resource.ResourceId) {
+				overlappingCount++
+			}
+		}
+	}
+	require.LessOrEqual(t, count, 310)
+	require.LessOrEqual(t, overlappingCount, 100)
```

Comment:
Can you clarify why there is overlap? As far as I can tell this is a single entrypoint in `view` which is `viewer` which has 410 tuples.

Is this because the cursor resumes from index `1` even though it streamed back index `1`?

Reply:
Removed